### PR TITLE
[Scheduling] Add CPSAT scheduler (via OR-Tools) for `SharedOperatorsProblem`

### PR DESCRIPTION
**Pull request by makslevental**
This a WIP implementation of a scheduler for the `SharedOperatorsProblem` using a CP-SAT solver, through OR-Tools.
The formulation/model is the [Resource Constrained Project Scheduling Problem](https://python-mip.readthedocs.io/en/latest/examples.html#:~:text=The%20Resource%2DConstrained%20Project%20Scheduling,required%20amount%20of%20different%20resources.) (but I do not use the formulation from that link). 

The current implementation is closely modeled on the [OR-Tools example](https://github.com/google/or-tools/blob/stable/examples/python/rcpsp_sat.py). That example is slightly more general (I think) than applies to CIRCT, namely it allows for multiple "recipes" for satisfying a task/op. Thus, the implementation needed to differ non-trivially. Since I am not a CP/SAT/scheduling/optimization wizard, my approach was to first build a parallel python version of that example, sans recipe functionality, all the while comparing the result of my altered script to the original. Hence all the auxilliary artifacts in `test/Scheduling/or-tools/cpsat_scripts/`; what you find there are problem instances ([kindly provided OR-Tools as well](https://github.com/google/or-tools/tree/stable/examples/data/rcpsp/single_mode)) along with a `test.sh` that runs OR-Tools python implementation and my python implementation and diffs the results. Once the python version was done, I one for one ported to C++/CIRCT (including naming conventions, to make the porting process more brainless).

I plan to normalize the naming (i.e., switch to using the CIRCT naming schemes) and remove all of the supporting material after in-tree tests are robust (enough coverage) and pass (which they currently do not).

Any and all comments/questions/concerns welcome.

cc @jopperm

### CODE REVIEW

**Comment by jopperm** [comment]
Awesome! Looks good on a very quick first glance. I'm currently away from my desk unfortunately, but I will do a detailed review early next week.

This should probably marked as a draft PR. For actually landing this later, you'll need to split the PR in smaller chunks (e.g. 1: boilerplate, test pass 2: actual implementation). Also, the supplementary scripts will help us review the algorithm, but should probably be dropped before landing as well.

**Review by jopperm — COMMENTED** [review]
Here's a first batch of questions/comments. Really cool that you're working on this -- helps the infra (and me...) to break out of my narrow LP-minded bubble 🤓

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (makslevental) [reply lib/Scheduling/CPSATSchedulers.cpp:56]: Yup I was confused by this as well but there's an overload for the [common case](https://github.com/google/or-tools/blob/49b6301e1e1e231d654d79b6032e79809868a70e/ortools/util/sorted_interval_list.h#L107).

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (makslevental) [reply lib/Scheduling/CPSATSchedulers.cpp:123]: I'll be honest: I do not know enough about optimization to make that call. Do the solvers behave the same in the presense of a variable with [0, inf) domain as they do in presense of that variable?

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (makslevental) [reply lib/Scheduling/CPSATSchedulers.cpp:56]: After offline (online on alternative channel?) I agree this is a bug. Thanks!

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (makslevental) [reply lib/Scheduling/CPSATSchedulers.cpp:127]: >The SharedOperatorsProblem assumes fully-pipelined operator instances. This means that the resource is used for only one time step (= an op's start time). I read this AddDemand call here as reservering the resource for the entire interval the operation is executing.

Interesting. That adds a bit of complexity and possibly makes this not the right problem class for this formulation, or said differently: the wrong formulation for this problem. What I mean is, if the operator is occupied only during the first interval, doesn't that mean the capacity is actually (maybe?) `pipeline depth * capacity`?

Comment (jopperm) [reply lib/Scheduling/CPSATSchedulers.cpp:127]: Do you need non-fully-pipelined operators for what you're trying to build? Defining a problem for that would be easy (just have to add an operator-type property "blockingTime"), and has been on my roadmap for quite some time, but I didn't have a use-case for it yet.

Anyways, yes, the formulation currently models a stricter problem than `SharedOperatorsProblem`. I'd expect that there are other approaches tailored to the fully-pipelined-operators special case. On the other hand, the formulation approach seems neat in general, and if there would be a way to set the blocking intervals as needed(1), this implementation could accept the `SharedOperatorsProblem` now, and a hypothetical `PartiallyPipelinedSharedOperatorsProblem`(2) later.

(1) Currently, you have:
```
interval_op = [startTime_op, startTime_op + latency]
```
Proposed for the `SharedOperatorProblem`:
```
interval_op = [startTime_op, startTime_op]
```
Proposed for hypothetical `PartiallyPipelinedSharedOperatorsProblem`:
```
interval_op = [startTime_op, startTime_op + blockingTime]
```

(2) Naming is hard.

**Comment by makslevental** [comment]
@jopperm sans renaming and comments this is ready for review.

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (makslevental) [reply lib/Scheduling/CPSATSchedulers.cpp]: I made changes such that the formulation conforms to the conventional `SharedOperatorProblem` and I'll put some thought into `PartiallyPipelinedSharedOperatorsProblem` soon.

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (makslevental) [inline lib/Scheduling/CPSATSchedulers.cpp:92]: conventional formulation for `SharedOperatorsProblem` (interval during which the resource is occupied has length `1`)

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (jopperm) [inline lib/Scheduling/CPSATSchedulers.cpp:84]: This type calls for an `auto` :wink:

Comment (jopperm) [reply lib/Scheduling/CPSATSchedulers.cpp:84]: Also, instead of iterating over all intervals over and over again, can't you just save the task intervals in a `Map<OperatorType, List<Interval>>`?

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (jopperm) [inline lib/Scheduling/CPSATSchedulers.cpp:87]: Plain `==` should be fine (operator types are `StringAttr`s).

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (jopperm) [inline lib/Scheduling/CPSATSchedulers.cpp:102]: Nit: `model.Minimize(makeSpan);` is clearer.

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (jopperm) [inline lib/Scheduling/CPSATSchedulers.cpp:109]: Nit: `auto *`, and omit the braces for a single-line loop.

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (jopperm) [inline lib/Scheduling/CPSATSchedulers.cpp:89]: `Twine(i)`, is that referencing the counter the variable construction loop?

**Review by jopperm — COMMENTED** [review]
Nice! Lots of nits (😇 ), but I appreciate how much simpler the formulation became after tailoring it to the `SharedOperatorsProblem`.

Please add a reference somewhere about the origins of the formulation (at least to the OR-Tools example that inspired you, or even better, a paper or textbook).

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (jopperm) [inline lib/Scheduling/CPSATSchedulers.cpp:93]: Please check if `const` is needed (here and below), it is at least uncommon in the code base to have local consts (though I'm not sure if it's a hard rule).

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (jopperm) [inline lib/Scheduling/CPSATSchedulers.cpp:106]: Micro-Nit: `Twine('_') + Twine(resource.strref())`

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (jopperm) [inline lib/Scheduling/CPSATSchedulers.cpp:96]: Efficiency nit: you could also not build and collect the intervals for unlimited resources in the first place.

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (jopperm) [reply lib/Scheduling/CPSATSchedulers.cpp:38]: (For the other schedulers, the rationale to get `lastOp` as an argument is that the client is expected to know which operation's start time it wants to minimize. In your case, this would save the constraints from all ops to the `makeSpan` variable. I'm fine with both approaches, just don't take an arg that you then ignore.)

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (jopperm) [inline lib/Scheduling/CPSATSchedulers.cpp:33]: Please add a reference here about the origins of the formulation (at least to the OR-Tools example that inspired you, or even better, a paper or textbook).

**Review by jopperm — COMMENTED** [review]
Thanks for iterating on this! Looks functionally correct to me now 👍 Left a few further remarks inline, but I think this can land very soon.

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (makslevental) [reply lib/Scheduling/CPSATSchedulers.cpp:38]: Right - your take on it is 100% accurate - prior to this iteration, I was buttressing `makeSpan` by `lastOp` but that wasn't working (for whatever reason) so I changed to having `makeSpan` cover all ops. Will drop this.

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (makslevental) [reply lib/Scheduling/CPSATSchedulers.cpp:74]: Interesting. Am I just misinformed or do `stl` maps/sets behave the same way (i.e. I've never needed to do this and somehow learned a weird habit)?

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (makslevental) [reply lib/Scheduling/CPSATSchedulers.cpp:106]: :D does the `.str()` incur a copy or something like that as opposed to the `strref()`? Probably.

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (makslevental) [reply lib/Scheduling/CPSATSchedulers.cpp:96]: So there's an issue here that revolves around `"barrier"(...) { opr = "_0", problemStartTime = 10 }` and `{ name = "_0", latency = 0 }` (i.e. no `limit = ...`). It's the issue that was tripping me [here](https://github.com/llvm/circt/pull/3350#discussion_r973097087). If I do

```cpp
  for (auto item : llvm::enumerate(tasks)) {
    ...
    auto resource = prob.getLinkedOperatorType(task);
    if (resource.hasValue()) {
      auto capacity = prob.getLimit(*resource);
      if (!capacity) {
        continue;
      }
      ...
      resourcesToTaskIntervals[resource.value()].emplace_back(taskInterval);
    }
  }
```

then neither the barriers nor the `return` op will be considered in the problem. The `return` will be handled by 

```cpp
cpModel.AddLessOrEqual(taskEnds[lastOp], makeSpan);
```

but the `"barrier"` op isn't *naturally* handled (I could of course include some string checking to catch `"barrier"`).

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (jopperm) [reply lib/Scheduling/CPSATSchedulers.cpp]: [std::map](https://en.cppreference.com/w/cpp/container/map/operator_at):
> operator[] is non-const because it inserts the key if it doesn't exist. If this behavior is undesirable or if the container is const, [at()](https://en.cppreference.com/w/cpp/container/map/at) may be used.

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (jopperm) [reply lib/Scheduling/CPSATSchedulers.cpp]: `.str()` builds an `std::string`, whereas `.strref()` returns some kind of reference to the data in the unique'd `StringAttr` (I think 😇 ). Again, a micro-nit :wink:

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (jopperm) [reply lib/Scheduling/CPSATSchedulers.cpp:109]: Ah ok, seems you have to at least _construct_ the interval variable in order to tie a task's start, end, and duration together (but I still think you don't have to store it in the map, as it only used again if the resource is limited).

Comment (jopperm) [reply lib/Scheduling/CPSATSchedulers.cpp:109]: > but the "barrier" op isn't naturally handled (I could of course include some string checking to catch "barrier").

No, don't do that, the (dummy!) barrier instruction from the test case shouldn't receive any special handling or mention in your code here.

Comment (makslevental) [reply lib/Scheduling/CPSATSchedulers.cpp]: > Ah ok, seems you have to at least _construct_ the interval variable in order to tie a task's start, end, and duration together (but I still think you don't have to store it in the map, as it only used again if the resource is limited).

This did the trick.

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (makslevental) [reply lib/Scheduling/CPSATSchedulers.cpp]: >Again, a micro-nit

The best kind :)

File: include/circt/Scheduling/Algorithms.h
Comment (jopperm) [inline include/circt/Scheduling/Algorithms.h:83]: This describes the wrong problem :wink:

May I propose:
```suggestion
/// Solve the acyclic problem with shared operators using constraint programming
/// and an external SAT solver. The objective is to minimize the start time of
/// the given \p lastOp. Fails if the dependence graph contains cycles, or \p
/// prob does not include \p lastOp.
```

The formulation *does* compute a schedule with the optimal makespan, correct?

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (jopperm) [inline lib/Scheduling/CPSATSchedulers.cpp:89]: Drop the "zero-distance" (here and a few lines below) please, that would only make sense in a cyclic problem.

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (jopperm) [inline lib/Scheduling/CPSATSchedulers.cpp:132]: Can't you just `cpModel.Minimize(taskEnds[lastOp])`?

**Review by jopperm — APPROVED** [review]
Awesome, this is good to go after addressing the documentation nits.

File: lib/Scheduling/CPSATSchedulers.cpp
Comment (makslevental) [reply lib/Scheduling/CPSATSchedulers.cpp]: yup

**Pull request merged**
